File: backend/python/routes/attendance.py
from flask import Blueprint, request, jsonify
from models import db, User, AttendanceSubject

attendance_bp = Blueprint('attendance', __name__)


def _get_user(reg_number):
    if not reg_number:
        return None
    return User.query.filter_by(registration_number=reg_number).first()


@attendance_bp.route('/list/<reg_number>', methods=['GET'])
def list_subjects(reg_number):
    user = _get_user(reg_number)
    if not user:
        return jsonify({"error": "User not found"}), 404

    subjects = AttendanceSubject.query.filter_by(user_id=user.id).order_by(AttendanceSubject.id).all()
    return jsonify([{
        "id": s.id,
        "name": s.name,
        "total": s.total,
        "attended": s.attended,
    } for s in subjects])


@attendance_bp.route('/subjects', methods=['POST'])
def add_subject():
    data = request.json or {}
    reg_number = data.get('registration_number')
    name = (data.get('name') or '').strip()

    if not reg_number or not name:
        return jsonify({"error": "Registration number and subject name are required"}), 400

    user = _get_user(reg_number)
    if not user:
        return jsonify({"error": "User not found"}), 404

    subject = AttendanceSubject(user_id=user.id, name=name, total=0, attended=0)
    db.session.add(subject)
    db.session.commit()

    return jsonify({
        "id": subject.id,
        "name": subject.name,
        "total": subject.total,
        "attended": subject.attended,
    }), 201


@attendance_bp.route('/subjects/<int:subject_id>', methods=['PATCH'])
def update_subject(subject_id):
    data = request.json or {}
    reg_number = data.get('registration_number')
    action = data.get('action')

    user = _get_user(reg_number)
    if not user:
        return jsonify({"error": "User not found"}), 404

    subject = AttendanceSubject.query.get(subject_id)
    if not subject or subject.user_id != user.id:
        return jsonify({"error": "Subject not found"}), 404

    if action == 'present':
        subject.attended += 1
        subject.total += 1
    elif action == 'absent':
        subject.total += 1
    else:
        return jsonify({"error": "Invalid action"}), 400

    db.session.commit()
    return jsonify({
        "id": subject.id,
        "name": subject.name,
        "total": subject.total,
        "attended": subject.attended,
    })


@attendance_bp.route('/subjects/<int:subject_id>', methods=['DELETE'])
def delete_subject(subject_id):
    reg_number = request.args.get('registration_number')
    user = _get_user(reg_number)
    if not user:
        return jsonify({"error": "User not found"}), 404

    subject = AttendanceSubject.query.get(subject_id)
    if not subject or subject.user_id != user.id:
        return jsonify({"error": "Subject not found"}), 404

    db.session.delete(subject)
    db.session.commit()
    return jsonify({"message": "Subject deleted"})


@attendance_bp.route('/reset/<reg_number>', methods=['DELETE'])
def reset_subjects(reg_number):
    user = _get_user(reg_number)
    if not user:
        return jsonify({"error": "User not found"}), 404

    AttendanceSubject.query.filter_by(user_id=user.id).delete()
    db.session.commit()
    return jsonify({"message": "Attendance reset"})

File: frontend/js/apps/attendance.js
class AttendanceApp extends BaseApp {
    constructor() {
        super("Attendance");
        this.apiBase = "/api/attendance";
        this.subjects = [];
        this.listContainer = null;
    }

    async loadSubjects() {
        const regNumber = this.getRegNumber();
        if (!regNumber) return false;

        try {
            const res = await fetch(`${this.apiBase}/list/${regNumber}`);
            const data = await res.json();
            if (!res.ok) throw new Error(data.error || "Failed to load subjects");
            this.subjects = Array.isArray(data) ? data : [];
            return true;
        } catch (err) {
            console.error("Attendance load failed:", err);
            alert(err.message || "Failed to load attendance data.");
            return false;
        }
    }

    render(container) {
        container.innerHTML = `
            <div class="attendance-container">
                <div class="attendance-header">
                    <h2>My Attendance</h2>
                    <div>
                        <button class="add-subject-btn">+ Add Subject</button>
                        <button class="reset-btn">⚠ Reset info</button>
                    </div>
                </div>
                <div class="subject-list">
                    <div style="color: #888; padding: 20px; text-align: center;">Loading subjects...</div>
                </div>
            </div>
        `;

        this.listContainer = container.querySelector(".subject-list");

        container.querySelector(".add-subject-btn").onclick = () => this.handleAddSubject();
        container.querySelector(".reset-btn").onclick = () => this.handleReset();

        this.loadSubjects().then((ok) => {
            if (ok) this.renderList();
        });
    }

    async handleAddSubject() {
        const regNumber = this.getRegNumber();
        if (!regNumber) return;

        const name = prompt("Subject Name:");
        if (!name || !name.trim()) return;

        try {
            const res = await fetch(`${this.apiBase}/subjects`, {
                method: 'POST',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ registration_number: regNumber, name: name.trim() })
            });
            const data = await res.json();
            if (!res.ok) throw new Error(data.error || "Failed to add subject");

            this.subjects.push(data);
            this.renderList();
        } catch (err) {
            alert(err.message || "Failed to add subject.");
        }
    }

    async handleReset() {
        const regNumber = this.getRegNumber();
        if (!regNumber) return;

        if (!confirm("Are you sure you want to reset all attendance data?")) return;

        try {
            const res = await fetch(`${this.apiBase}/reset/${regNumber}`, { method: 'DELETE' });
            const data = await res.json();
            if (!res.ok) throw new Error(data.error || "Failed to reset");

            this.subjects = [];
            this.renderList();
        } catch (err) {
            alert(err.message || "Failed to reset attendance.");
        }
    }

    async updateSubject(subjectId, action) {
        const regNumber = this.getRegNumber();
        if (!regNumber) return;

        try {
            const res = await fetch(`${this.apiBase}/subjects/${subjectId}`, {
                method: 'PATCH',
                headers: { 'Content-Type': 'application/json' },
                body: JSON.stringify({ registration_number: regNumber, action })
            });
            const data = await res.json();
            if (!res.ok) throw new Error(data.error || "Failed to update");

            const index = this.subjects.findIndex(s => s.id === subjectId);
            if (index !== -1) this.subjects[index] = data;
            this.renderList();
        } catch (err) {
            alert(err.message || "Failed to update attendance.");
        }
    }

    async deleteSubject(subjectId) {
        const regNumber = this.getRegNumber();
        if (!regNumber) return;

        try {
            const res = await fetch(`${this.apiBase}/subjects/${subjectId}?registration_number=${encodeURIComponent(regNumber)}`, {
                method: 'DELETE'
            });
            const data = await res.json();
            if (!res.ok) throw new Error(data.error || "Failed to delete");

            this.subjects = this.subjects.filter(s => s.id !== subjectId);
            this.renderList();
        } catch (err) {
            alert(err.message || "Failed to delete subject.");
        }
    }

    renderList() {
        if (!this.listContainer) return;

        this.listContainer.innerHTML = "";

        if (this.subjects.length === 0) {
            this.listContainer.innerHTML = `
                <div style="color: #888; padding: 20px; text-align: center;">
                    No subjects yet. Click "+ Add Subject" to get started.
                </div>`;
            return;
        }

        this.subjects.forEach((sub) => {
            const percent = sub.total === 0 ? 0 : Math.round((sub.attended / sub.total) * 100);
            const statusClass = percent >= 75 ? "safe" : "danger";
            const bunkMsg = this.calculateBunk(sub.attended, sub.total);

            const el = document.createElement("div");
            el.className = `subject-card ${statusClass}`;
            el.innerHTML = `
                <div class="sub-info">
                    <h3>${sub.name}</h3>
                    <div class="sub-stats">
                        <span>${sub.attended}/${sub.total}</span>
                        <span class="percent">${percent}%</span>
                    </div>
                    <div class="bunk-status">${bunkMsg}</div>
                </div>
                <div class="sub-actions">
                    <button class="present-btn">✅</button>
                    <button class="absent-btn">❌</button>
                    <button class="delete-btn" style="background: transparent; border: 1px solid #ccc;">🗑</button>
                </div>
            `;

            el.querySelector(".present-btn").onclick = () => this.updateSubject(sub.id, 'present');
            el.querySelector(".absent-btn").onclick = () => this.updateSubject(sub.id, 'absent');
            el.querySelector(".delete-btn").onclick = () => {
                if (confirm(`Delete ${sub.name}?`)) {
                    this.deleteSubject(sub.id);
                }
            };

            this.listContainer.appendChild(el);
        });
    }

    calculateBunk(attended, total) {
        const target = 0.75;
        const current = total === 0 ? 0 : attended / total;

        if (current >= target) {
            const canBunk = Math.floor((attended - target * total) / target);
            return canBunk > 0 ? `You can bunk ${canBunk} classes` : "On the edge!";
        }

        const need = Math.ceil((target * total - attended) / (1 - target));
        return `Attend ${need} next classes!`;
    }
}

EventBus.on("SYSTEM_BOOT", () => {
    const launcher = document.querySelector("[data-app='attendance']");
    if (launcher) {
        launcher.onclick = () => {
            WindowManager.createWindow(new AttendanceApp());
        };
    }
});
